File: backend/src/nausicass_global_green_initiative_api/services/monitoring/sentry_monitoring.py
from typing import Any, Optional

import sentry_sdk
from sentry_sdk.integrations.flask import FlaskIntegration
from sentry_sdk.integrations.sqlalchemy import SqlalchemyIntegration

from .monitoring_interface import MonitoringService, MonitoringTransaction
from .sentry_transaction import SentryTransaction


class SentryMonitoringService(MonitoringService):
    """
    Sentry implementation of the monitoring interface.
    """

    def init(self, **kwargs: Any) -> None:
        sentry_sdk.init(
            dsn=kwargs["dsn"],
            environment=kwargs.get("environment", "production"),
            release=kwargs.get("release"),
            traces_sample_rate=kwargs.get("traces_sample_rate", 1.0),
            profiles_sample_rate=kwargs.get("profiles_sample_rate", 0.1),
            send_default_pii=True,
            integrations=[
                FlaskIntegration(
                    transaction_style="url",
                ),
                SqlalchemyIntegration(),
            ],
            # Capture all request bodies for max detail
            max_request_body_size="always",
            # Attach server name for tracing
            server_name=kwargs.get("server_name"),
            debug=kwargs.get("debug", False),
        )

    def capture_exception(
        self, error: BaseException, context: Optional[dict] = None
    ) -> None:
        with sentry_sdk.push_scope() as scope:
            if context:
                for key, value in context.items():
                    scope.set_extra(key, value)
            sentry_sdk.capture_exception(error)

    def capture_message(self, message: str, level: str = "info") -> None:
        sentry_sdk.capture_message(message, level=level)

    def set_user(self, user: Optional[dict]) -> None:
        sentry_sdk.set_user(user)

    def add_breadcrumb(
        self,
        category: str,
        message: str,
        level: str = "info",
        data: Optional[dict] = None,
    ) -> None:
        sentry_sdk.add_breadcrumb(
            category=category,
            message=message,
            level=level,
            data=data or {},
        )

    def start_transaction(
        self, name: str, op: str, **kwargs: Any
    ) -> MonitoringTransaction:
        return SentryTransaction(name=name, op=op, **kwargs)

    def set_tag(self, key: str, value: str) -> None:
        sentry_sdk.set_tag(key, value)

    def set_context(self, name: str, context: dict) -> None:
        sentry_sdk.set_context(name, context)

File: frontend/src/services/monitoring/sentry/sentryMonitoring.ts
import React from "react"
import * as Sentry from "@sentry/react"
import {
  useLocation,
  useNavigationType,
  createRoutesFromChildren,
  matchRoutes,
} from "react-router-dom"

import type {
  Breadcrumb,
  MonitoringService,
  MonitoringUser,
  SeverityLevel,
} from "../types"
import { SentryTransaction } from "./sentryTransaction"

export type SentryConfig = {
  dsn: string
  environment: string
  release?: string
  tracesSampleRate?: number
  profileSessionSampleRate?: number
  replaysSessionSampleRate?: number
  replaysOnErrorSampleRate?: number
  debug?: boolean
}

function toSentrySeverity(
  level: SeverityLevel,
): "fatal" | "error" | "warning" | "info" | "debug" {
  return level
}

// Factory function that creates
// Sentry-based monitoring service
export function createSentryMonitoringService(
  config: SentryConfig,
): MonitoringService {
  Sentry.init({
    dsn: config.dsn,
    environment: config.environment,
    release: config.release,
    integrations: [
      Sentry.reactRouterV7BrowserTracingIntegration({
        useEffect: React.useEffect,
        useLocation,
        useNavigationType,
        createRoutesFromChildren,
        matchRoutes,
      }),
      Sentry.browserProfilingIntegration(),
      Sentry.replayIntegration({
        maskAllText: false,
        maskAllInputs: true,
        blockAllMedia: false,
      }),
      Sentry.breadcrumbsIntegration({
        console: true,
        dom: true,
        fetch: true,
        history: true,
        xhr: true,
      }),
      Sentry.httpClientIntegration(),
    ],
    tracesSampleRate: config.tracesSampleRate ?? 1.0,
    profileSessionSampleRate: config.profileSessionSampleRate ?? 1.0,
    replaysSessionSampleRate: config.replaysSessionSampleRate ?? 0.1,
    replaysOnErrorSampleRate: config.replaysOnErrorSampleRate ?? 1.0,
    sendDefaultPii: false,
    debug: config.debug ?? false,
  })

  return {
    // Capture exception with optional context
    captureException(error: unknown, context?: Record<string, unknown>): void {
      if (context) {
        Sentry.captureException(error, { extra: context })
      } else {
        Sentry.captureException(error)
      }
    },

    // Capture message with optional severity level
    captureMessage(message: string, level?: SeverityLevel): void {
      Sentry.captureMessage(
        message,
        level ? toSentrySeverity(level) : undefined,
      )
    },

    // Set user context for monitoring
    setUser(user: MonitoringUser | null): void {
      Sentry.setUser(user)
    },

    // Add a breadcrumb to Sentry with optional level and data
    addBreadcrumb(breadcrumb: Breadcrumb): void {
      const crumb: Sentry.Breadcrumb = {
        category: breadcrumb.category,
        message: breadcrumb.message,
      }
      if (breadcrumb.level) crumb.level = toSentrySeverity(breadcrumb.level)
      if (breadcrumb.data) crumb.data = breadcrumb.data
      Sentry.addBreadcrumb(crumb)
    },

    // Start a transaction with the given context
    startTransaction(context) {
      return new SentryTransaction(context)
    },

    // Set a tag in Sentry for additional context
    setTag(key: string, value: string): void {
      Sentry.setTag(key, value)
    },

    // Set context in Sentry
    setContext(name: string, context: Record<string, unknown>): void {
      Sentry.setContext(name, context)
    },
  }
}
